llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.cpp

```diff
@@ -28,6 +31,11 @@ void SMEAttrs::set(unsigned M, bool Enable) {
          "ZA_New and ZA_NoLazySave are mutually exclusive");
   assert(!(hasSharedZAInterface() && (Bitmask & ZA_NoLazySave)) &&
          "ZA_Shared and ZA_NoLazySave are mutually exclusive");
+
+  // ZT0 Attrs
+  assert((!sharesZT0() || (hasNewZT0Body() ^ isZT0In() ^ isZT0InOut() ^
+                           isZT0Out() ^ preservesZT0())) &&
+         "ZT0_New,In,Out,InOut,Preserved are mutually exclusive");
```

Comment:
Is it possible to tighten this up and make it more specific? First, I think it should have full names like `ZT0_New, ZT0_In, ZT0_Out` and have spaces after the commas to make it a bit easier to read. Second, sorry if I'm sending you on a wild goose chase, is it possible to do something along the lines of

```c++
SmallVector<StringRef> Attrs;
if (hasNewZT0Body())
  Attrs.push_back("ZT0_New");
if (isZT0In())
  Attrs.push_back("ZT0_In");
// ...
// test other mutually exclusive attributes

if (Attrs.size() > 1){
  // join Attrs into a string and bail as an llvm_unreachable
}
``` 
?
That way we print a better diagnostic since we know exactly what mutually exclusive attributes are causing the problem

Reply:
This is not really a diagnostic, but rather a message of an assert. Asserts aren't really user-facing, they are more a way to identify an issue, which has to be clear enough to understand what the issue is. From the message, we know that mutually exclusive attributes have been combined somehow, which is enough to start debugging and find the cause of it.

I agree with your comment about having spaces between the commas and having the full names.

Reply:
I've reworded this assert to use the full names, it's now:
`ZT0_New, ZT0_In, ZT0_Out, ZT0_InOut and ZT0_Preserved are all mutually exclusive`